packages/expo-modules-core/android/src/main/java/expo/modules/core/utilities/EmulatorUtilities.kt

```diff
@@ -0,0 +1,33 @@
+package expo.modules.core.utilities
+
+import android.os.Build
+import java.util.*
+
+object EmulatorUtilities {
+  // Adapted from https://github.com/react-native-device-info/react-native-device-info/blob/ea9f868a80acaec68583094c891098a03ecb411a/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java#L225
+  fun isRunningOnEmulator(): Boolean {
+    return Build.FINGERPRINT.startsWith("generic") ||
+      Build.FINGERPRINT.startsWith("unknown") ||
+      Build.MODEL.contains("google_sdk") ||
+      Build.MODEL.lowercase(Locale.ROOT).contains("droid4x") ||
+      Build.MODEL.contains("Emulator") ||
+      Build.MODEL.contains("Android SDK built for x86") ||
+      Build.MANUFACTURER.contains("Genymotion") ||
+      Build.HARDWARE.contains("goldfish") ||
+      Build.HARDWARE.contains("ranchu") ||
+      Build.HARDWARE.contains("vbox86") ||
+      Build.PRODUCT.contains("sdk") ||
```

Comment:
i was thinking the rule for product name containing *sdk* is not very strict here. well forget it since the code is originally referenced from flutter 😅